### PR TITLE
FIX: Re-add cohort identifier to template name

### LINKED ISSUE #415: fmriprep error "ValueError: 'MNIPediatricAsym:cohort-3' is not in list"
### What happened?

I am trying to run fmriprep for preprocessing of resting state fmri data of a pediatric population with a mean age of 7.5. I would like to use the MNIPediatricAsym:cohort-3 template included in templateflow as output space for the spatial normalization of the T1w images, however I keep getting the error that this template is not in list (please see log output). 

The $TEMPLATEFLOW_HOME environment is set to "/home/em23/.cache/templateflow" and includes the "tpl-MNIPediatricAsym" directory, which contains 6 cohort directories, license, scripts directory, and the template_description.json file. The cohort directories all contain nii.gz files for brain mask, CSF_probseg, GM_probseg, WM_probseg, PD, T1w, and T2w for both res-1 and res-2 (14 files per cohort). 

I also ran fmriprep with the MNI152NLin2009cAsym template as the output space (this template is the default, so same command line used with --output-spaces flag omitted), which did not produce any errors. 

Has anyone encountered this problem before while running fmriprep for a pediatric population? Many thanks in advance!

### What command did you use?

```shell
fmriprep-docker home/em23/Desktop/fmriprep home/em23/Desktop/fmriprep_output participant --skip_bids_validation --participant-label 93 --fs-license-file usr/local/freesurfer/license.txt --ignore slicetiming --output-spaces MNIPediatricAsym:cohort-3
```

### What version of the software are you running?

fMRIPrep v23.2.0

### How are you running this software?

Docker

### Is your data BIDS valid?

No

### Are you reusing any previously computed results?

FreeSurfer

### Please copy and paste any relevant log output.

```shell
Errors
Node Name: fmriprep_23_2_wf.sub_93_wf.template_iterator_wf.select_xfm

File: /out/sub-93/log/20240216-164851_5103920b-fdb1-4480-b5ae-d361c2f2d229/crash-20240216-191305-root-select_xfm.a0-65ff9fe2-92bf-4e29-acdc-f7f1c812bb60.txt
Working Directory: /tmp/work/fmriprep_23_2_wf/sub_93_wf/template_iterator_wf/_in_tuple_MNIPediatricAsymcohort-3.cohort3.resnative/select_xfm
Inputs:

    anat2std_xfm: ['/out/sub-93/anat/sub-93_from-T1w_to-MNIPediatricAsym_mode-image_xfm.h5', '/out/sub-93/anat/sub-93_from-T1w_to-MNI152NLin2009cAsym_mode-image_xfm.h5']
    key: MNIPediatricAsym:cohort-3
    keys: ['MNIPediatricAsym', 'MNI152NLin2009cAsym']

Traceback (most recent call last):
  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/nipype/pipeline/plugins/multiproc.py", line 344, in _send_procs_to_workers
    self.procs[jobid].run(updatehash=updatehash)
  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/nipype/pipeline/engine/nodes.py", line 527, in run
    result = self._run_interface(execute=True)
  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/nipype/pipeline/engine/nodes.py", line 645, in _run_interface
    return self._run_command(execute)
  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/nipype/pipeline/engine/nodes.py", line 771, in _run_command
    raise NodeExecutionError(msg)
nipype.pipeline.engine.nodes.NodeExecutionError: Exception raised while executing Node select_xfm.

Traceback:
	Traceback (most recent call last):
	  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/nipype/interfaces/base/core.py", line 400, in run
	    outputs = self.aggregate_outputs(runtime)
	  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/nipype/interfaces/base/core.py", line 429, in aggregate_outputs
	    predicted_outputs = self._list_outputs()  # Predictions from _list_outputs
	  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/niworkflows/interfaces/utility.py", line 217, in _list_outputs
	    index = self.inputs.keys.index(self.inputs.key)
	ValueError: 'MNIPediatricAsym:cohort-3' is not in list



Node Name: fmriprep_23_2_wf.sub_93_wf.anat_fit_wf.anat_reports_wf.template_iterator_wf.select_xfm

File: /out/sub-93/log/20240216-164851_5103920b-fdb1-4480-b5ae-d361c2f2d229/crash-20240216-191305-root-select_xfm.a0-c177ea7c-8cc6-468a-b53b-f3bc559eb63a.txt
Working Directory: /tmp/work/fmriprep_23_2_wf/sub_93_wf/anat_fit_wf/anat_reports_wf/template_iterator_wf/_in_tuple_MNIPediatricAsymcohort-3.cohort3.resnative/select_xfm
Inputs:

    anat2std_xfm: ['/out/sub-93/anat/sub-93_from-T1w_to-MNIPediatricAsym_mode-image_xfm.h5', '/out/sub-93/anat/sub-93_from-T1w_to-MNI152NLin2009cAsym_mode-image_xfm.h5']
    key: MNIPediatricAsym:cohort-3
    keys: ['MNIPediatricAsym', 'MNI152NLin2009cAsym']

Traceback (most recent call last):
  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/nipype/pipeline/plugins/multiproc.py", line 344, in _send_procs_to_workers
    self.procs[jobid].run(updatehash=updatehash)
  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/nipype/pipeline/engine/nodes.py", line 527, in run
    result = self._run_interface(execute=True)
  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/nipype/pipeline/engine/nodes.py", line 645, in _run_interface
    return self._run_command(execute)
  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/nipype/pipeline/engine/nodes.py", line 771, in _run_command
    raise NodeExecutionError(msg)
nipype.pipeline.engine.nodes.NodeExecutionError: Exception raised while executing Node select_xfm.

Traceback:
	Traceback (most recent call last):
	  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/nipype/interfaces/base/core.py", line 400, in run
	    outputs = self.aggregate_outputs(runtime)
	  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/nipype/interfaces/base/core.py", line 429, in aggregate_outputs
	    predicted_outputs = self._list_outputs()  # Predictions from _list_outputs
	  File "/opt/conda/envs/fmriprep/lib/python3.10/site-packages/niworkflows/interfaces/utility.py", line 217, in _list_outputs
	    index = self.inputs.keys.index(self.inputs.key)
	ValueError: 'MNIPediatricAsym:cohort-3' is not in list
```

### Additional information / screenshots

_No response_

### PR DESCRIPTION
**Pull request by mgxd**
Fixes #415 

We have been stripping out the cohort identifier in `split_desc`, and as a result losing the information when saving out transforms. This re-enables the output transforms to include the cohort information, i.e. `...from-MNIPediatricAsym+3_to-T1w_mode-image_xfm.h5`.

However, this change exposes another problem: `collect_derivatives` cannot collect these transformations, as the current `BIDSLayout` configurations do not account for the `+` character during parsing. Since BIDS derivatives conventions are still a bit murky, I figured we want to just create a modified `fmriprep-derivatives` config to catch this convention when parsing.

### CODE REVIEW

**Comment by oesteban** [comment]
Looks good in general. Perhaps relying on the `+` semantic is too risky. Since the proposal was done also including `-` as allowed (and I believe also underscore), it will not fly anytime soon. And would honestly not will to try to get only `+` accepted.

What do you think about `from-MNIPediatricAsymCohort3` or, alternatively and potentially generating conflicts down the line, `from-MNIPediatricAsym_to-T1w_desc-cohort3_xfm.h5`?

**Comment by mgxd** [comment]
TBH I don't love either option - at first I figured we could just add the `cohort-` entity directly, but that will clobber cases of inter-cohort xfms. IMO the `+` seems to be the cleanest / least ambiguous...

**Comment by mgxd** [comment]
@effigies if you have a minute, would appreciate a review

File: pyproject.toml [hunk 1]
```diff
@@ -25,7 +25,8 @@ dependencies = [
     "matplotlib >= 2.2.0",
     "nibabel >= 4.0.1",
     "nipype >= 1.7.0",
-    "niworkflows >= 1.8.0",
+#    "niworkflows >= 1.8.0",
+    "niworkflows @ git+https://github.com/nipreps/niworkflows.git@master",
```
Comment (effigies) [inline pyproject.toml:29]:
```suggestion
    "niworkflows @ git+https://github.com/nipreps/niworkflows.git@master",
```

**Review by effigies — APPROVED** [review]
Mostly LGTM.

**Comment by oesteban** [comment]
> TBH I don't love either option - at first I figured we could just add the `cohort-` entity directly, but that will clobber cases of inter-cohort xfms. IMO the `+` seems to be the cleanest / least ambiguous...

In that case, I believe `space-` could be used: `space-MNIPediatricAsym_from-cohort1_to-cohort2_xfm.h5`, but this honestly feels like a templateflow's issue, not a plausible output of sMRIPrep

File: smriprep/utils/tests/__init__.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+from pathlib import Path
+
+DERIV_SKELETON = Path(__file__).parent / ('derivatives.yml')
```
Comment (effigies) [inline smriprep/utils/tests/__init__.py:3]: Woops. Didn't click the green button on this one. Anyway, after all the effort purging `__file__`, I think we should not re-introduce it...

```suggestion
from niworkflows.data import Loader

load_data = Loader(__package__)

DERIV_SKELETON = load_data('derivatives.yml')
```

**Comment by effigies** [comment]
Is this considered a bugfix or a feature?

**Comment by mgxd** [comment]
I consider it more of a bug fix to get `cohort` processing / `--derivatives` working, as the changes will really only affect those cases. But this will invalidate the cache starting from the `fit.registration` workflow so maybe worth a minor bump?

**Comment by mgxd** [comment]
Given that we're almost in Q2 of 2024 let's cut 24.0

**Pull request merged**